internal/gatewayapi/listeners_test.go

```diff
@@ -23,339 +23,493 @@ import (
-			Reason:  string(gatewayapi_v1beta1.ListenerReasonInvalid),
-			Message: "invalid hostname \".invalid.$.\": [a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]",
-		},
-	}, res.InvalidListenerConditions)
+		res := ValidateListeners(listeners)
+		assert.Empty(t, res.Ports)
+		assert.Equal(t, map[gatewayapi_v1beta1.SectionName]metav1.Condition{
+			"listener-1": {
+				Type:    string(gatewayapi_v1beta1.ListenerConditionProgrammed),
+				Status:  metav1.ConditionFalse,
+				Reason:  string(gatewayapi_v1beta1.ListenerReasonInvalid),
+				Message: "invalid hostname \"192.168.1.1\": must be a DNS name, not an IP address",
+			},
+			"listener-2": {
+				Type:    string(gatewayapi_v1beta1.ListenerConditionProgrammed),
+				Status:  metav1.ConditionFalse,
+				Reason:  string(gatewayapi_v1beta1.ListenerReasonInvalid),
+				Message: "invalid hostname \"*.*.projectcontour.io\": [a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\\*\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]",
+			},
+			"listener-3": {
+				Type:    string(gatewayapi_v1beta1.ListenerConditionProgrammed),
+				Status:  metav1.ConditionFalse,
+				Reason:  string(gatewayapi_v1beta1.ListenerReasonInvalid),
+				Message: "invalid hostname \".invalid.$.\": [a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]",
+			},
+		}, res.InvalidListenerConditions)
+	})
+
+	t.Run("Three HTTPS listeners on the same port, each with a different hostname", func(t *testing.T) {
+		listeners := []gatewayapi_v1beta1.Listener{
+			{
+				Name:     "https-1",
+				Protocol: gatewayapi_v1beta1.HTTPSProtocolType,
+				Hostname: ref.To(gatewayapi_v1beta1.Hostname("https-1.gateway.projectcontour.io")),
+				Port:     443,
+			},
+			{
+				Name:     "https-2",
+				Protocol: gatewayapi_v1beta1.HTTPSProtocolType,
+				Hostname: ref.To(gatewayapi_v1beta1.Hostname("https-2.gateway.projectcontour.io")),
+				Port:     443,
+			},
+			{
+				Name:     "https-3",
+				Protocol: gatewayapi_v1beta1.HTTPSProtocolType,
+				Hostname: ref.To(gatewayapi_v1beta1.Hostname("https-3.gateway.projectcontour.io")),
+				Port:     443,
+			},
+		}
+
+		res := ValidateListeners(listeners)
+		assert.Len(t, res.InvalidListenerConditions, 0)
+		assert.Len(t, res.Ports, 1)
+		assert.Len(t, res.ListenerNames, 3)
+	})
+
+	t.Run("Conflicting protocols on a port", func(t *testing.T) {
```

Comment:
i dont know if this is covered elsewhere but seems like our new custom protocol could cause a conflict with a HTTP port as well

Reply:
implementation above looks correct, but a test case to make sure we don't cause any issues in the future would be great!